components/script/dom/servoparser/async_html.rs

```diff
@@ -20,218 +21,249 @@ use dom::htmltemplateelement::HTMLTemplateElement;
+            html_tokenizer_sender: to_html_tokenizer_sender,
+            sink_receiver: from_sink_receiver,
+            sink_sender: to_sink_sender,
             nodes: HashMap::new(),
-            document_node: ParseNode {
-                id: 0,
-                qual_name: None,
-            }
+            url: url
         };
-        let data = ParseNodeData::default();
-        sink.insert_parse_node_data(0, data);
-        sink.insert_node(0, JS::from_ref(document.upcast()));
-        sink
+        tokenizer.insert_node(0, JS::from_ref(document.upcast()));
+
+        let mut sink = Sink::new(sink_sender, sink_receiver);
+        let mut ctxt_parse_node = None;
+        let mut form_parse_node = None;
+        let mut fragment_context_is_some = false;
+        if let Some(fc) = fragment_context {
+            let node = sink.new_parse_node();
+            tokenizer.insert_node(node.id, JS::from_ref(fc.context_elem));
+            ctxt_parse_node = Some(node);
+
+            form_parse_node = fc.form_elem.map(|form_elem| {
+                let node = sink.new_parse_node();
+                tokenizer.insert_node(node.id, JS::from_ref(form_elem));
+                node
+            });
+            fragment_context_is_some = true;
+        };
+
+        // Create new thread for HtmlTokenizer. This is where parser actions
+        // will be generated from the input provided. These parser actions are then passed
+        // onto the main thread to be executed.
+        thread::Builder::new().name(String::from("HTML Parser")).spawn(move || {
+            run(sink,
+                fragment_context_is_some,
+                ctxt_parse_node,
+                form_parse_node,
+                html_tokenizer_sender,
+                html_tokenizer_receiver);
+        }).expect("HTML Parser thread spawning failed");
+
+        tokenizer
     }
 
-    fn new_parse_node(&mut self) -> ParseNode {
-        let id = self.next_parse_node_id.get();
-        let data = ParseNodeData::default();
-        self.insert_parse_node_data(id, data);
-        self.next_parse_node_id.set(id + 1);
-        ParseNode {
-            id: id,
-            qual_name: None,
+    #[allow(unsafe_code)]
+    pub fn feed(&mut self, input: &mut BufferQueue) -> Result<(), Root<HTMLScriptElement>> {
+        enum Message {
+            Sink(FromSinkMsg),
+            HtmlTokenizer(FromHtmlTokenizerMsg)
+        }
+
+        let mut send_tendrils = Vec::new();
+        while let Some(str) = input.pop_front() {
+            send_tendrils.push(SendTendril::from(str));
+        }
+        let send_tendrils = Arc::new(Mutex::new(send_tendrils));
+        // Send message to parser thread, asking it to started reading from the shared input.
+        // Parser operation messages will be sent to main thread as they are evaluated.
+        self.html_tokenizer_sender.send(ToHtmlTokenizerMsg::Feed(send_tendrils.clone())).unwrap();
```

Comment:
Why do we need to put the tendrils in a `Arc<Mutex<T>>` if we just send them afterwards?